Add compile time instruction count metric

laithsakka opened the pull request:
PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133834



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

laithsakka committed:
add compile time instruction count metric

[ghstack-poisoned]

laithsakka committed:
Update on "add compile time instruction count metric"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "add compile time instruction count metric"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
address lint error

laithsakka committed:
Update on "add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
```
update_hint_regression,instruction_count,11538293410
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
```
update_hint_regression,instruction_count,11538293410
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py:
```diff
@@ -34,12 +34,15 @@ def f(a, b):
 def main():
     result_path = sys.argv[1]
-    Benchmark().enable_instruction_count().collect_all().append_results(result_path)
+    Benchmark().enable_instruction_count().enable_compile_time_instruction_count().collect_all().append_results(
```
Are you really proposing that when we have dozens of benchmark we have to manually update each of them to record new metrics we think we want to record?

laithsakka replied:
i assume those benchmarks are going to be added one by one? and not all metrics need to be enabled on all of them. 
if someone is adding  a dozen of benchmarks and want to enable all metrics on all of them then he can easily create a function (enableAll()).

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -3067,3 +3068,42 @@ def get_user_object_from_id(obj_id):
+    if config.record_compile_time_instruction_count:
+
+        @functools.wraps(func)
+        def wrapper_function(*args, **kwargs):
```
No decorator please, it makes the cProfiles worse. Do this as a context manager.

laithsakka replied:
I can do that.

ezyang reviewed (changes requested):
see comments

ezyang commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -2,47 +2,69 @@
 
+    def enable_compile_time_instruction_count(self):
+        self._compile_time_instruction_count = True
+        return self
```
At this point I think it's good to actually add docs to all of these. In particular, it seems good to carefully define what compile time instruction count means, because, e.g., if I do a forward-backward compile, I'm pretty sure your current formulation doesn't capture backwards compilation.

laithsakka replied:
i will be very interesting if it does not because that means we do not capture that in our c_profiles nor out strobelight profiler, because   @maybe_cprofile is only wrapping _compile_inner

laithsakka replied:
I will add a documentation though

laithsakka replied:
is there a specific entry point for the missing compilation part that i can wrap?

ezyang commented on benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py:
```diff
@@ -34,12 +34,15 @@ def f(a, b):
 
-        f(self.input, self.splits)
+        for i in range(1000):
+            f(self.input, self.splits)
```
If you want to be in the business of measuring the instruction count of compiled codes, it seems like it would be better to do a measurement here that *excludes* compilation time from the perf measurement. The most typical way to do this would be to do the warmup (forcing compilation) and then do the perf measurement in steady-state. Otherwise, your overall instruction count number is going to bounce up and down, mostly due to changes in compile time, not so useful for finding runtime regressions unless you take care to subtract out the compile time.

laithsakka replied:
the instruction count metric, measure the total instruction count for the work function, this could be something that does not even do compilation . (people can add micro benchmarks). 

I see your point about adding another metric which is " instruction count of compiled codes" but that would be a different diff. 

for the time being i can exclude the enable_instruction_count() for the update hint benchmark since what we care about for that one is compile time instruction count. 

I was using that as basic benchmark example, maybe i will add a dummy benchmark were i enable everything on it as example.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -3067,3 +3068,42 @@ def get_user_object_from_id(obj_id):
+    _id: Optional[int] = None
+
+    @classmethod
+    def start(cls):
```
nit: -> None on this and all the other class methods

laithsakka committed:
Update on "add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
```
update_hint_regression,instruction_count,11538293410
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
- add type annotation for class methods return 
- document each metric in benchmark base. 
- only collect compile time metric for update_hint benchmark. 
- avoid using decorator.

laithsakka committed:
Update on "add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
```
update_hint_regression,instruction_count,11538293410
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
```
update_hint_regression,instruction_count,11538293410
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
- after rebase i noticed the logging to scribe happening from benchmark base, I added a todo, will follow up on this once we update the logger.

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented:
I'm on vacation next week, feel free to get someone else to review/approve/land, I don't have any blockers, just haven't done a rereview in time.

aorenste commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -51,44 +53,68 @@
-    def count_instructions(self):
+    def _count_instructions(self):
         print(f"collecting instruction count for {self.name()}")
-        self.prepare_once()
```
Do we not need this call anymore? Oh - I guess it moved to collect_all()...

laithsakka replied:
yeh once for all metrics not needed

aorenste commented on torch/_dynamo/utils.py:
```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
+
+    def __enter__(self):
+        if config.record_compile_time_instruction_count:
+            self.started = True
```
Do we want this to be able to handle recursive calls?
```
with CompileTimeInstructionCollector():
  ...
  with CompileTimeInstructionCollector():
    ...
```
Is so might want a counter instead of a bool...

laithsakka replied:
right now its not used with recursions but i can make it resilinet to that

aorenste commented on torch/_dynamo/utils.py:
```diff
@@ -56,6 +56,7 @@
 from typing_extensions import Literal, TypeGuard
 
 import torch
+import torch._C._instruction_counter as i_counter
```
We only use this twice in this file - I wouldn't bother with such a short name. Maybe just do the import without an alias and call `_instruction_counter.blah` directly?

aorenste commented on torch/_dynamo/utils.py:
```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
+        return cls._counter
+
+
+class CompileTimeInstructionCollector:
```
This would probably be easier to read as a contextmanager...

laithsakka replied:
you mean something like
```
from contextlib import contextmanager

@contextmanager
def closing(thing):
    try:
        yield thing
    finally:
        thing.close()
 ```

aorenste replied:
yes

aorenste commented on torch/_dynamo/utils.py:
```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
+    _id: int = -1
+
+    @classmethod
+    def start(cls) -> None:
```
Maybe annotate all these functions?

laithsakka replied:
what do you mean

aorenste replied:
Oh - I'm being silly here. @classmethod cls doesn't need annotation - but below the
```
def __init__(self):
```
should be:
```
def __init__(self) -> None:
```
(and enter and exit need type annotations if you don't switch to a contextmanager.

aorenste commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -51,44 +53,68 @@
 
+    # measure total number of instruction spent in convert_frame.compile_inner
+    # TODO is there other parts we need to add ?
+    _compile_time_instruction_count = False
```
This name is a bit ambiguous (is it a enable? Is it a value?) maybe something like `_enable_compile_time_instruction_count`? (Same issue w/ `_instruction_count` above)

aorenste commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -51,44 +53,68 @@
+            self._work()
+            count = CompileTimeInstructionCounter.value()
+            print(f"compile time instruction count for iteration {i} is {count}")
+            if i != 0:
```
Why skip 0? Warmup? Since we're returning `min` warmup shouldn't matter... Please add a comment.

laithsakka replied:
yeh it does not really matter we can remove it

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

aorenste commented on torch/_dynamo/utils.py:
```diff
@@ -56,6 +56,7 @@
 from typing_extensions import Literal, TypeGuard
 
 import torch
+import torch._C._instruction_counter as _instruction_counter
```
nit: this is usually written as:
```
from torch._C import _instruction_counter
```

aorenste commented on torch/_dynamo/utils.py:
```diff
@@ -3106,3 +3107,39 @@ def get_user_object_from_id(obj_id):
     user_obj_id_to_weakref[obj_id] = weakref.ref(obj)
+
+
+class CompileTimeInstructionCounter:
```
Why is this "CompileTimeInstructionCounter" and not just "CpuInstructionCounter"? It has no actual relationship to compile time, does it? (Other than we intend to use it for such)

laithsakka replied:
This should not be used around anything else but counting compile time instructions since its accessed globally 
as CompileTimeInstructionCounter.value()

laithsakka replied:
if we call it CpuInstructionCounter we can create an a global instance and 
compileTimeInstructionCounter = CpuInstructionCounter()
but i dont see right now a reason for generalizing it

aorenste commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -64,29 +75,42 @@ def description(self):
+            results.append(count)
+        return min(results)
+
+    def _count_compile_time_instructions(self):
```
I'm confused as to the difference between `_count_instructions` and `_count_compile_time_instructions`.  Don't they do the same thing but one uses _instruction_counter directly and the other indirectly?

laithsakka replied:
# measure total number of instruction spent in _work.
    _enable_instruction_count = False

    # measure total number of instruction spent in convert_frame.compile_inner
    # TODO is there other parts we need to add ?
    _enable_compile_time_instruction_count = False

laithsakka replied:
# measure total number of instruction spent in _work.
 _enable_instruction_count = False 
 
 # measure total number of instruction spent in convert_frame.compile_inner # TODO is there other parts we need to add ? _enable_compile_time_instruction_count = False

laithsakka replied:
if your program  is a micro beanchmark measure some function in isolate. you would use _enable_instruction_count and not _enable_compile_time_instruction_count because
_enable_compile_time_instruction_count would return 0.

aorenste commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -64,29 +75,42 @@ def description(self):
+
+    def _count_compile_time_instructions(self):
+        print(f"collecting compile time instruction count for {self.name()}")
+        config.record_compile_time_instruction_count = True
```
Add a quick comment here noting that because inner_compile (or whatever) calls record() this causes it to actually record the inner_compile ticks and not the whole thing.

aorenste reviewed (approved):
A couple nits but LGTM

laithsakka commented on benchmarks/dynamo/pr_time_benchmarks/benchmark_base.py:
```diff
@@ -64,29 +75,42 @@ def description(self):
+
+    def _count_compile_time_instructions(self):
+        print(f"collecting compile time instruction count for {self.name()}")
+        config.record_compile_time_instruction_count = True
```
i will add that, but in general if anything throws here the binary will exit anyway.

laithsakka replied:
actually what are you suggesting what we do in the finally? 
            CompileTimeInstructionCounter.clear() ?
we always clear anyway before work()

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
@pytorchbot  merge

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

laithsakka commented:
@pytorchbot  merge

laithsakka committed:
Update on "Add compile time instruction count metric"

 PYTHONPATH=$(pwd) python benchmarks/update_hint_benchmark.py out
as of this diff, compile_time_instruction_count counts the number of instruction from within
convert_frame.compile_inner
```
update_hint_regression,compile_time_instruction_count,10522459165
```
 will add result from CI once populated.                                               



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

laithsakka commented:
@pytorchbot  merge
